cpp/src/arrow/dataset/dataset.cc

```diff
@@ -99,9 +103,42 @@ FragmentIterator Source::GetFragments(std::shared_ptr<ScanOptions> scan_options)
   return GetFragmentsImpl(std::move(simplified_scan_options));
 }
 
+struct VectorRecordBatchGenerator {
+  explicit VectorRecordBatchGenerator(std::vector<std::shared_ptr<RecordBatch>> batches)
+      : batches_(std::move(batches)) {}
+
+  RecordBatchIterator operator()() const { return MakeVectorIterator(batches_); }
+
+  std::vector<std::shared_ptr<RecordBatch>> batches_;
+};
+
+InMemorySource::InMemorySource(std::shared_ptr<Schema> schema,
+                               std::vector<std::shared_ptr<RecordBatch>> batches)
+    : Source(std::move(schema)),
+      get_batches_(VectorRecordBatchGenerator(std::move(batches))) {}
+
+InMemorySource::InMemorySource(std::shared_ptr<Table> table)
+    : Source(table->schema()), get_batches_([table] {
+        auto reader = std::make_shared<TableBatchReader>(*table);
+        return MakeFunctionIterator([reader, table] { return reader->Next(); });
+      }) {}
+
 FragmentIterator InMemorySource::GetFragmentsImpl(
     std::shared_ptr<ScanOptions> scan_options) {
-  return MakeVectorIterator(fragments_);
+  auto create_batch =
+      [scan_options](std::shared_ptr<RecordBatch> batch) -> std::shared_ptr<Fragment> {
+    std::vector<std::shared_ptr<RecordBatch>> batches;
+
+    while (batch->num_rows() > scan_options->batch_size) {
+      batches.push_back(batch->Slice(0, scan_options->batch_size));
+      batch = batch->Slice(scan_options->batch_size);
```

Comment:
This leads to a lot of intermediary unused batches, i.e. n-1 precisely. This adds a lot of pressure to all the shared_ptr objects (types, array, array_data, ...).

Reply:
Okay

Reply:
While you're at it, it might be worth extracting this in a small utility, I feel like we're going to need it elsewhere.